xls/contrib/integrator/integration_algorithms/basic_integration_algorithm.h

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// A abstract class for an algorithm to integrate multiple functions.
+
+#ifndef XLS_INTEGRATOR_BASIC_INTEGRATION_ALGORITHM_
+#define XLS_INTEGRATOR_BASIC_INTEGRATION_ALGORITHM_
+
+#include "xls/contrib/integrator/integration_algorithms/integration_algorithm.h"
+
+namespace xls {
+
+// A naive merging algorithm.  A node is eligible to be added to the
+// integration function when all of its operands have already been added.
+// At each step, adds the eligible node for which the cost of adding it to
+// the function (either by inserting or merging with any integeration function
+// node) is the lowest.
+class BasicIntegrationAlgorithm
+    : public IntegrationAlgorithm<BasicIntegrationAlgorithm> {
```

Comment:
Is CRTP necessary? Why doesn't normal inheritance work?

Reply:
We'd like to call a single static function to run an algorithm (which is some child class of IntegrationAlgorithm) to completion and return the result.  This is in contrast to a 2-step process where we 1) construct an instance of an algorithm then 2) run the algorithm (more prone to misuse).  XLS uses a similar general strategy (without inheritance) elsewhere (e.g. code generators).  

The issue is that the static function that we call belongs to the parent IntegrationAlgorithm class, but must call the constructor of a child class algorithm.  Currently, I don't know another way of doing this other than CRTP.  However, CRTP is akward to deal with (and advantage of single call vs. build + run calls isn't huge), so I'm open to suggestions for other approaches.

Reply:
That's fine. Worth a comment describing the motivation here.